array.c

```diff
@@ -3126,20 +3127,59 @@ rb_ary_to_a(VALUE ary)
 
 /*
  *  call-seq:
- *     ary.to_h                  -> hash
- *     ary.to_h {|item| block }  -> hash
+ *    array.to_h -> new_hash
+ *    array.to_h {|item| ... } -> new_hash
+ *
+ *  Returns a new \Hash formed from +self+.
+ *
+ *  In all the forms below, each 2-element \Array
+ *  is formed into a key-value pair in the new \Hash.
+ *
+ *  ====== \Array of 2-Element Arrays
```

Comment:
This should be restructured.  The first section should discuss usage with a block, in which case the block is called and should return a 2-element array.  The second section should state if a block is not given, the array must already be composed of 2-element arrays, which are used to construct the hash.  This reduces the need to explain 4 cases (2-element array with block, 2-element array without block, other array with block, other array with block), to 3 cases (block, array of 2-element arrays without block, other array without block), which makes more sense as the behavior is the same for all cases where a block is passed.

Reply:
Agree that block-first is much better.   But doesn't the block get a different  yield for 2-ele and other?

Reply:
In your example code for both cases, you yield the element in the array (`item`).  In the array of 2-element arrays case, it's a 2-element array, and in the other case it is some other object, but the behavior is the same in both cases.

Reply:
That insight came to me during lunch (BLT, pinot grigio).  Thanks.

Reply:
Fixed.